pkg/agent/multicluster/gateway_route_controller.go

```diff
@@ -0,0 +1,343 @@
+		namespace:            namespace,
+	}
+	controller.gwInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueueGateway(cur)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				controller.enqueueGateway(cur)
+			},
+			DeleteFunc: func(old interface{}) {
+				controller.enqueueGateway(old)
+			},
+		},
+		resyncPeriod,
+	)
+	controller.ciImportInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueueClusterInfoImport(cur)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				controller.enqueueClusterInfoImport(cur)
+			},
+			DeleteFunc: func(old interface{}) {
+				controller.enqueueClusterInfoImport(old)
+			},
+		},
+		resyncPeriod,
+	)
+	return controller
+}
+
+func (c *GatewayController) enqueueGateway(obj interface{}) {
+	err := checkGateway(obj, c.queue)
+	if err != nil {
+		return
+	}
+}
+
+func (c *GatewayController) enqueueClusterInfoImport(obj interface{}) {
+	err := checkClusterInfoImport(obj, c.queue)
+	if err != nil {
+		return
+	}
+}
+
+// Run will create defaultWorkers workers (go routines) which will process
+// the Gateway events from the workqueue.
+func (c *GatewayController) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s", gwcontrollerName)
+	defer klog.Infof("Shutting down %s", gwcontrollerName)
+	cacheSyncs := []cache.InformerSynced{c.gwListerSynced, c.ciImportListerSynced}
+	if !cache.WaitForNamedCacheSync(gwcontrollerName, stopCh, cacheSyncs...) {
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+// worker is a long-running function that will continually call the processNextWorkItem
+// function in order to read and process a message on the workqueue.
+func (c *GatewayController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *GatewayController) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	if k, ok := obj.(key); !ok {
+		c.queue.Forget(obj)
+		klog.Errorf("Expected type 'key' in work queue but got object", "object", obj)
+		return true
+	} else if err := c.syncMCRouteOnGatewayNode(k); err == nil {
+		c.queue.Forget(k)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(k)
+		klog.ErrorS(err, "Error syncing key, requeuing.", "key", k)
+	}
+	return true
+}
+
+func (c *GatewayController) syncMCRouteOnGatewayNode(k key) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).InfoS("Finished syncing Gateway Node route for Multi-cluster.", "type", k.kind,
+			"name", k.name, "time", time.Since(startTime))
+	}()
+	switch k.kind {
+	case "Gateway":
```

Comment:
The case block is indeed long. Maybe better to define two funcs for the two cases.